src/main/java/net/nextencia/rrdiagram/Server.java

```diff
@@ -0,0 +1,245 @@
+  }
+}
+
+class BNFHandler implements HttpHandler {
+
+  Logger logger = Logger.getLogger(BNFHandler.class.getName());
+
+  @Override
+  public void handle(HttpExchange httpExchange) throws IOException {
+    if("GET".equals(httpExchange.getRequestMethod())) { 
+      handleRequest(httpExchange);
+    }
+  }
+
+  public Map<String, String> parseParameters(HttpExchange exchange) {
+    URI requestedUri = exchange.getRequestURI();
+    String query = requestedUri.getRawQuery();
+    Map<String, String> parameters = new HashMap<String, String>();
+     if (query != null) {
+       String pairs[] = query.split("[&]");
+
+       for (String pair : pairs) {
+        String param[] = pair.split("[=]");
+
+        String key = null;
+        String value = null;
+        if (param.length > 0) {
+          try {
+            key = URLDecoder.decode(param[0],System.getProperty("file.encoding"));
+          } catch (java.io.UnsupportedEncodingException ue) {
+            ue.printStackTrace();
+          }
+        }
+
+        if (param.length > 1) {
+          try {
+            value = URLDecoder.decode(param[1],System.getProperty("file.encoding"));
+          } catch (java.io.UnsupportedEncodingException ue) {
+            ue.printStackTrace();
+          }
+        }
+        
+        if (key != null && value != null) {
+          parameters.put(key, value);
+        }
+      }
+    }
+    return parameters;
+  }
+
+  public String cleanString(String input) {
+    // whitespaces
+    input = input.replaceAll("\\s*", "");
+    // leading ,
+    input = input.replaceAll("^,+","");
+    // trailing ,
+    input = input.replaceAll(",+$","");
+    // multiple ,
+    input = input.replaceAll(",+",",");
+    return input;
+  }
+
+  public void handleRequest(HttpExchange httpExchange) {
+    int responseCode = HttpCodes.OK;
+    try {
+      // parse params
+      Map<String, String> params = parseParameters(httpExchange);
+
+      BNFProcessor bnfprocessor = BNFProcessor.get(params.get("api"), params.get("version"));
+      String content = "";
+      
+      if (bnfprocessor != null) {
+        // fetch the data
+        if (params.containsKey("mode")) {
+          if (params.get("mode").equals("grammar") || params.get("mode").equals("diagram")) {
+            String strrules[] = null;
+            String localrefs[] = null;
+            if (params.containsKey("rules")) {
+              strrules = params.get("rules").split(",");
+            }
+
+            if (params.containsKey("local")) {
+              localrefs = cleanString(params.get("local")).split(",");
+            }
+
+            int depth=0;
+            if (params.containsKey("depth")) {
+              try {
+                depth = Integer.parseInt(params.get("depth"));
+              } catch (NumberFormatException nfe) {
+                logger.warning("invalid depth:" + params.get("depth"));
+              }
+            }
```

Comment:
it should error on invalid depth, not attempt to continue with depth 0.

create a helper function that returns content in the good case or error in the bad case
- good case: serve content, 200
- bad case: serve error message, 4xx
could change up this idea, such as also logging the error message and/or returning generic error response

Reply:
done.